### PR TITLE
Explain the difference between now() and mz_logical_timestamp()

### PR DESCRIPTION
**Pull request by msudberg**
Also added some incidental formatting corrections to SELECT doc.

### CODE REVIEW

File: doc/user/content/sql/functions/now_and_mz_logical_timestamp.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+title: "now and mz_logical_timestamp functions"
+description: "In Materialize mz_logical_timestamp() acts the way now() does in most systems."
+menu:
+  main:
+    parent: 'sql-functions'
+---
+
+In Materialize, `now()` doesn't represent the system time; it represents the time with timezone when the query was executed. By contrast, `mz_logical_timestamp()` represents the logical time at which the query executes.
+
+In most Materialize queries, you should use `mz_logical_timestamp()` where you might use `now()` in other systems, for example, when defining [temporal filters](https://materialize.com/temporal-filters/) or when creating materialized views from multiple sources with potentially conflicting system times.
```
Comment (mjibson) [inline doc/user/content/sql/functions/now_and_mz_logical_timestamp.md:11]: > or when creating materialized views from multiple sources

I don't think we allow `mz_logical_timestamp` in views unless it's in a temporal filter? If that's correct, this phrase should maybe be removed.

File: doc/user/content/sql/functions/now_and_mz_logical_timestamp.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+SELECT content, insert_ts, delete_ts
+FROM events
+WHERE mz_logical_timestamp() >= insert_ts
+  AND mz_logical_timestamp()  < delete_ts;
+```
+
+Next, you'll populate the table with timestamp data.
+
+```sql
+insert into events VALUES (
```
Comment (awang) [inline doc/user/content/sql/functions/now_and_mz_logical_timestamp.md:35]: nit: if we're capitalizing the CREATEs above, should we also capitalize INSERT INTO here (to keep it consistent)?

**Review by awang — APPROVED** [review]
can we add an example showing how to do a temporal filter relative to now()? 
basically, what matt describes here: https://materializeinc.slack.com/archives/CUFFL8UBC/p1618591034076200

```
CREATE MATERIALIZED VIEW kinesis_view2 AS
   SELECT extract(epoch from split_part(text, e'\t', 3)::timestamp) * 1000 
   FROM kinesis_source 
   WHERE (extract(epoch from split_part(text, e'\t', 3)::timestamp) * 1000)::bigint + 86400000 >= mz_logical_timestamp() 
```

otherwise lgtm

**Comment by mjibson** [comment]
`mz_logical_timestamp` currently has 3 different meanings:

1. instructing mz to produce a temporal filter
2. asking mz the system time at which a one-off query is run
3. wall clock now in INSERTs (although I am attempting to remove this in #6732, so this bullet can hopefully be ignored)

AFAIK it cannot be used in other places. For example it you can't query it in views, it's only valid in WHERE clauses to create temporal filters:

```
materialize=> create view v as select mz_logical_timestamp() from t;
ERROR:  mz_logical_timestamp cannot be used in static queries
```

Currently this document is a nice summary of how to write a temporal filter using it, but ignores the original and still more common use of `mz_logical_timestamp()` which is as a debugging tool for us to figure out the system time at which a query will run. I think it's very hard to explain what system time in materialize is because it's 1) not a global thing, 2) is determined by each source and 3) sometimes has nothing to do with seconds or the Unix epoch because it's just a counter that starts at 1 and goes up. We should perhaps avoid discussing this stuff because it's very tricky to get right and be useful to users?

I think if we are going to document how this function differs from `now`, we need to explain these two modes of operation, because they are very different. Maybe something like:

> `mz_logical_timestamp()` is a function that deals with materialized's system time. It can be used to filter data by its system time as a temporal filter as described in this blog post `[link]`. It can also be used in one-off statements to see the system time at which a query will execute. This can be useful if you need to understand the time up to which data has been ingested by materialized, and corresponds to the `timestamp` column of `TAIL` `[link]`. Generally a `CREATE SOURCE` will ingest data and produce timestamps that correspond to "milliseconds since the Unix epoch".

I might propose that instead of trying to explain how now and mz_logical_timestamp are different (because doing it correctly would involve a discussion about system time), this page could be like "how to use mz_logical_timestamp in a temporal filter"? I also have no idea what the motivation for this page is though, so maybe that's a bad idea for reasons I don't know.

**Comment by msudberg** [comment]
@mjibson - Thank you, this was very helpful! I will rework and resubmit.

The motivation is that users have been attempting to use `now()` the way they do for other systems and have not understood the results, so I think we do have to mention it. But it can be more of an aside and less of the main focus.

**Comment by mjibson** [comment]
Other things we can consider: undocumenting `mz_logical_timestamp` in the functions page. Or changing its text to something like "used for temporal filters and internal debugging" to further sway people away from using it.

**Comment by msudberg** [comment]
> can we add an example showing how to do a temporal filter relative to now()?

@awang - I added an example of a query using now() against timestamps, but I don't think you can use it in a temporal filter? At least, none of my attempts work, but I may be doing something wrong.

**Comment by msudberg** [comment]
@mjibson, could you take another look at this to see if it makes sense now? Thank you!

File: doc/user/data/sql_funcs.yml [hunk 3]
```diff
@@ -376,12 +376,14 @@
     url: extract
 
   - signature: mz_logical_timestamp() -> numeric
-    description: 'The logical time at which a query executes.<br/><br/>**NOTE**: Users
+    description: 'The logical time at which a query executes. Used for temporal filters and internal debugging.<br/><br/>**NOTE**: Users
```
Comment (mjibson) [inline doc/user/data/sql_funcs.yml:379]: The NOTE here isn't accurate. You can use this function in views, as long as its part of a temporal filter.

File: doc/user/content/sql/select.md [hunk 4]
```diff
@@ -123,6 +123,16 @@ queries' semantics.
 
 For an example, see [Using CTEs](#using-ctes).
 
+#### Known limitations
```
Comment (mjibson) [inline doc/user/content/sql/select.md:126]: Is this stuff supposed to be here? Feels like it should be in another PR.

File: doc/user/content/sql/functions/now_and_mz_logical_timestamp.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
+---
+title: "now and mz_logical_timestamp functions"
+description: "In Materialize mz_logical_timestamp() acts the way now() does in most systems."
+menu:
+  main:
+    parent: 'sql-functions'
+---
+
+In Materialize, `now()` doesn't represent the system time, as it does in most systems; it represents the time with timezone when the query was executed. `mz_logical_timestamp()` comes closer to what `now()` typically indicates, since it represents the logical time at which the query executes, based on the system time defined by the system on which `materialized` is installed. Neither `now()` nor `mz_logical_timestamp()` can be used when creating views.
```
Comment (mjibson) [inline doc/user/content/sql/functions/now_and_mz_logical_timestamp.md:9]: mz_logical_timestamp can be used in a view in a temporal filter.

File: doc/user/content/sql/functions/now_and_mz_logical_timestamp.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+The typical uses for `mz_logical_timestamp()` are:
+
+* Internal debugging
+* Defining [temporal filters](https://materialize.com/temporal-filters/) (`now` can also be used for this purpose)
```
Comment (mjibson) [inline doc/user/content/sql/functions/now_and_mz_logical_timestamp.md:14]: `now` can't be used for temporal filters

File: doc/user/content/sql/functions/now_and_mz_logical_timestamp.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+WHERE mz_logical_timestamp() >= insert_ts
+  AND mz_logical_timestamp()  < delete_ts;
+```
+
+Next, you'll populate the table with timestamp data.
+
+```sql
+INSERT INTO events VALUES (
+    'hello',
+    mz_logical_timestamp(),
```
Comment (mjibson) [inline doc/user/content/sql/functions/now_and_mz_logical_timestamp.md:50]: I just shipped a PR that prevents use of `mz_logical_timestamp` in INSERT: https://github.com/MaterializeInc/materialize/pull/6732. You should be able to use `extract(epoch from now()) * 1000` instead.
